serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,288 @@
+package client
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-hclog"
+)
+
+// maxRetries is the maximum number of times the client
+// should retry.
+const maxRetries = 10
+
+var (
+	ErrNamespaceUnset = errors.New(`"namespace" is unset`)
+	ErrPodNameUnset   = errors.New(`"podName" is unset`)
+	ErrNotInCluster   = errors.New("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined")
+)
+
+// New instantiates a Client. The stopCh is used for exiting retry loops
+// when closed.
+func New(logger hclog.Logger, stopCh <-chan struct{}) (*Client, error) {
+	config, err := inClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		logger: logger,
+		config: config,
+		stopCh: stopCh,
+	}, nil
+}
+
+// Client is a minimal Kubernetes client.
```

Comment:
It might be worth commenting here why we elected to not bring an existing client.

Reply:
Oh, good idea, will do.

Reply:
This is done in https://github.com/hashicorp/vault/pull/8249/commits/c116c5b57f5b17c0087aea3190d6187a6148eda0